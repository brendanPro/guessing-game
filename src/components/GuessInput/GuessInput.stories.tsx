import "@/index.css"
import type { Meta, StoryObj } from '@storybook/react';
import { GuessInput } from './GuessInput';
import { fn } from '@storybook/test';

const meta = {
  title: 'Game/GuessInput',
  component: GuessInput,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    onSubmit: {
      action: 'submitted',
      description: 'Callback when a guess is submitted',
    },
    disabled: {
      control: 'boolean',
      description: 'Whether the input is disabled',
    },
    maxLength: {
      control: { type: 'number', min: 1, max: 20 },
      description: 'Maximum length of the input',
    },
  },
  args: {
    onSubmit: fn(),
  },
} satisfies Meta<typeof GuessInput>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    disabled: false,
    maxLength: 7,
  },
};

export const ShortName: Story = {
  args: {
    disabled: false,
    maxLength: 3,
  },
};

export const MediumName: Story = {
  args: {
    disabled: false,
    maxLength: 8,
  },
};

export const LongName: Story = {
  args: {
    disabled: false,
    maxLength: 12,
  },
};

export const Disabled: Story = {
  args: {
    disabled: true,
    maxLength: 7,
  },
};

export const NoMaxLength: Story = {
  args: {
    disabled: false,
    maxLength: undefined,
  },
};

export const Interactive: Story = {
  args: {
    disabled: false,
    maxLength: 7,
  },
  render: (args) => (
    <div className="space-y-4">
      <div className="text-center">
        <p className="text-sm text-muted-foreground mb-2">
          Try typing "pikachu" (7 letters) to enable the submit button
        </p>
        <GuessInput {...args} />
      </div>
    </div>
  ),
};

export const WithDifferentMaxLengths: Story = {
  args: {
    disabled: false,
    onSubmit: fn(),
  },
  render: (args) => (
    <div className="space-y-4 w-full max-w-md">
      <div>
        <p className="text-sm font-semibold mb-2">Max Length: 3 (e.g., "Mew")</p>
        <GuessInput {...args} maxLength={3} />
      </div>
      <div>
        <p className="text-sm font-semibold mb-2">Max Length: 7 (e.g., "Pikachu")</p>
        <GuessInput {...args} maxLength={7} />
      </div>
      <div>
        <p className="text-sm font-semibold mb-2">Max Length: 12 (e.g., "Tortank-MÃ©ga")</p>
        <GuessInput {...args} maxLength={12} />
      </div>
    </div>
  ),
};

export const DisabledState: Story = {
  args: {
    disabled: true,
    maxLength: 7,
  },
  render: (args) => (
    <div className="space-y-4">
      <div className="text-center">
        <p className="text-sm text-muted-foreground mb-2">
          Input is disabled (game over or loading)
        </p>
        <GuessInput {...args} />
      </div>
    </div>
  ),
};

