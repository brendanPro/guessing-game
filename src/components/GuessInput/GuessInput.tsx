import React, { useState, useEffect, useRef } from 'react';
import { Button } from '@/components/ui/button';
import { Input } from '@/components/ui/input';
import { Card, CardContent } from '@/components/ui/card';
import { cn } from '@/lib/utils';

interface GuessInputProps {
  onSubmit: (guess: string) => void;
  disabled?: boolean;
  maxLength?: number;
  className?: string;
}

export function GuessInput({ onSubmit, disabled = false, maxLength, className }: GuessInputProps) {
  const [guess, setGuess] = useState('');
  const inputRef = useRef<HTMLInputElement>(null);

  // Auto-focus the input when component mounts or when not disabled
  useEffect(() => {
    if (!disabled && inputRef.current) {
      inputRef.current.focus();
    }
  }, [disabled]);

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (guess.trim() && !disabled && maxLength && guess.length === maxLength) {
      onSubmit(guess.trim());
      setGuess('');
      // Re-focus after submission
      setTimeout(() => {
        if (inputRef.current && !disabled) {
          inputRef.current.focus();
        }
      }, 100);
    }
  };

  const handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      handleSubmit(e);
    }
  };

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    if (!maxLength || value.length <= maxLength) {
      setGuess(value);
    }
  };

  return (
    <Card className={cn("w-full max-w-md mx-auto", className)}>
      <CardContent className="p-3 sm:p-4">
        <form onSubmit={handleSubmit} className="flex gap-2">
          <Input
            ref={inputRef}
            type="text"
            value={guess}
            onChange={handleInputChange}
            onKeyPress={handleKeyPress}
            placeholder="Entrez le nom du Pokemon..."
            disabled={disabled}
            className="flex-1"
            autoComplete="off"
            maxLength={maxLength}
          />
          <Button 
            type="submit" 
            disabled={!guess.trim() || disabled || !maxLength || guess.length !== maxLength}
            variant="default"
          >
            Deviner
          </Button>
        </form>
      </CardContent>
    </Card>
  );
}
