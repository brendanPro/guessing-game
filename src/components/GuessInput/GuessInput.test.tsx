import { describe, test, expect, mock } from "bun:test";
import { renderToStaticMarkup } from "react-dom/server";
import { GuessInput } from "./GuessInput";


describe("GuessInput Component", () => {
  const mockOnSubmit = mock(() => {});

  test("snapshot with default props", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toMatchSnapshot();
  });

  test("snapshot with short maxLength", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={3} disabled={false} />
    );
    expect(html).toMatchSnapshot();
  });

  test("snapshot with long maxLength", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={12} disabled={false} />
    );
    expect(html).toMatchSnapshot();
  });

  test("snapshot when disabled", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={true} />
    );
    expect(html).toMatchSnapshot();
  });

  test("snapshot without maxLength", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} disabled={false} />
    );
    expect(html).toMatchSnapshot();
  });

  test("contains input field", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('<input');
  });

  test("contains submit button", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('Deviner');
  });

  test("has correct placeholder text", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('Entrez le nom du Pokemon...');
  });

  test("input has type text", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('type="text"');
  });

  test("input has autocomplete off", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('autoComplete="off"');
  });

  test("input has maxLength attribute when provided", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('maxLength="7"');
  });

  test("button is disabled by default (empty input)", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('disabled=""');
  });

  test("input is disabled when disabled prop is true", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={true} />
    );
    expect(html).toContain('disabled=""');
  });

  test("form has correct structure", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('<form');
    expect(html).toContain('</form>');
  });

  test("has flex gap layout", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('gap-2');
  });

  test("input has flex-1 class", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('flex-1');
  });

  test("button has type submit", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('type="submit"');
  });

  test("applies custom className when provided", () => {
    const html = renderToStaticMarkup(
      <GuessInput 
        onSubmit={mockOnSubmit} 
        maxLength={7} 
        disabled={false}
        className="custom-class"
      />
    );
    expect(html).toContain('custom-class');
  });

  test("has responsive padding classes", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={7} disabled={false} />
    );
    expect(html).toContain('p-3');
    expect(html).toContain('sm:p-4');
  });

  test("maxLength of 3 sets correct attribute", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={3} disabled={false} />
    );
    expect(html).toContain('maxLength="3"');
  });

  test("maxLength of 12 sets correct attribute", () => {
    const html = renderToStaticMarkup(
      <GuessInput onSubmit={mockOnSubmit} maxLength={12} disabled={false} />
    );
    expect(html).toContain('maxLength="12"');
  });
});

