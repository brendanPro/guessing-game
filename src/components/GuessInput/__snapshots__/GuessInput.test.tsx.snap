// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`GuessInput Component snapshot with default props 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><form class="flex gap-2"><input type="text" data-slot="input" class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 aria-invalid:outline-destructive/60 dark:aria-invalid:outline-destructive dark:aria-invalid:ring-destructive/40 aria-invalid:ring-destructive/20 aria-invalid:border-destructive/60 dark:aria-invalid:border-destructive flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:ring-4 focus-visible:outline-1 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-[3px] aria-invalid:focus-visible:outline-none md:text-sm dark:aria-invalid:focus-visible:ring-4 flex-1" placeholder="Entrez le nom du Pokemon..." autoComplete="off" maxLength="7" value=""/><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-9 px-4 py-2 has-[&gt;svg]:px-3" type="submit" disabled="">Deviner</button></form></div></div>"`;

exports[`GuessInput Component snapshot with short maxLength 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><form class="flex gap-2"><input type="text" data-slot="input" class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 aria-invalid:outline-destructive/60 dark:aria-invalid:outline-destructive dark:aria-invalid:ring-destructive/40 aria-invalid:ring-destructive/20 aria-invalid:border-destructive/60 dark:aria-invalid:border-destructive flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:ring-4 focus-visible:outline-1 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-[3px] aria-invalid:focus-visible:outline-none md:text-sm dark:aria-invalid:focus-visible:ring-4 flex-1" placeholder="Entrez le nom du Pokemon..." autoComplete="off" maxLength="3" value=""/><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-9 px-4 py-2 has-[&gt;svg]:px-3" type="submit" disabled="">Deviner</button></form></div></div>"`;

exports[`GuessInput Component snapshot with long maxLength 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><form class="flex gap-2"><input type="text" data-slot="input" class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 aria-invalid:outline-destructive/60 dark:aria-invalid:outline-destructive dark:aria-invalid:ring-destructive/40 aria-invalid:ring-destructive/20 aria-invalid:border-destructive/60 dark:aria-invalid:border-destructive flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:ring-4 focus-visible:outline-1 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-[3px] aria-invalid:focus-visible:outline-none md:text-sm dark:aria-invalid:focus-visible:ring-4 flex-1" placeholder="Entrez le nom du Pokemon..." autoComplete="off" maxLength="12" value=""/><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-9 px-4 py-2 has-[&gt;svg]:px-3" type="submit" disabled="">Deviner</button></form></div></div>"`;

exports[`GuessInput Component snapshot when disabled 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><form class="flex gap-2"><input type="text" data-slot="input" class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 aria-invalid:outline-destructive/60 dark:aria-invalid:outline-destructive dark:aria-invalid:ring-destructive/40 aria-invalid:ring-destructive/20 aria-invalid:border-destructive/60 dark:aria-invalid:border-destructive flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:ring-4 focus-visible:outline-1 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-[3px] aria-invalid:focus-visible:outline-none md:text-sm dark:aria-invalid:focus-visible:ring-4 flex-1" placeholder="Entrez le nom du Pokemon..." disabled="" autoComplete="off" maxLength="7" value=""/><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-9 px-4 py-2 has-[&gt;svg]:px-3" type="submit" disabled="">Deviner</button></form></div></div>"`;

exports[`GuessInput Component snapshot without maxLength 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><form class="flex gap-2"><input type="text" data-slot="input" class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 aria-invalid:outline-destructive/60 dark:aria-invalid:outline-destructive dark:aria-invalid:ring-destructive/40 aria-invalid:ring-destructive/20 aria-invalid:border-destructive/60 dark:aria-invalid:border-destructive flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:ring-4 focus-visible:outline-1 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 aria-invalid:focus-visible:ring-[3px] aria-invalid:focus-visible:outline-none md:text-sm dark:aria-invalid:focus-visible:ring-4 flex-1" placeholder="Entrez le nom du Pokemon..." autoComplete="off" value=""/><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-9 px-4 py-2 has-[&gt;svg]:px-3" type="submit" disabled="">Deviner</button></form></div></div>"`;
